File: problems/j/jewelrybox/jewelrybox.py
#! /usr/bin/python3
"""
Author: Rudi César Comiotto Modena
Email: rudi.modena@gmail.com
"""
from math import sqrt


def calc_roots_derivative(X, Y):
    if X < Y:
        threshold = X / 2
    else:
        threshold = Y / 2

    a = 12
    b = (-4 * X) + (-4 * Y)
    c = X * Y
    delta = b ** 2 - 4 * a * c
    root_1 = ((- 1 * b) + sqrt(delta)) / (2 * a)
    root_2 = ((- 1 * b) - sqrt(delta)) / (2 * a)

    list_valid_roots = list()
    if (0 < root_1) and (root_1 < threshold):
        list_valid_roots.append(root_1)

    if (0 < root_2) and (root_2 < threshold):
        list_valid_roots.append(root_2)

    return list_valid_roots


def calc_volume(list_valid_roots, X, Y):
    biggest_volume = 0
    for h in list_valid_roots:
        volume = (4 * h ** 3) - (2 * X * h ** 2) - (2 * Y * h ** 2) + (X * Y * h)
        if volume > biggest_volume:
            biggest_volume = volume

    return biggest_volume


def main():
    t = int(input())

    for _ in range(t):
        X, Y = [int(val) for val in input().split()]
        list_valid_roots = calc_roots_derivative(X, Y)
        print(calc_volume(list_valid_roots, X, Y))


if __name__ == '__main__':
    main()

File: problems/j/jewelrybox/jewelrybox.cpp
/*
Author: Rudi César Comiotto Modena
Email: rudi.modena@gmail.com
*/

#include <bits/stdc++.h>

using namespace std;

double calc_root_derivative(int X, int Y) {
    int threshold;

    if (X < Y) {
        threshold = X / 2.0;
    }
    else {
        threshold = Y / 2.0;
    }

    int a = 12.0;
    int b = (-4.0 * X) + (-4.0 * Y);
    int c = X * Y;
    double delta = pow(b, 2.0) - 4.0 * a * c;
    double root_1 = ((- 1.0 * b) + sqrt(delta)) / (2.0 * a);
    double root_2 = ((- 1.0 * b) - sqrt(delta)) / (2.0 * a);  

    if (0 < root_1 && root_1 < threshold) {
        return root_1;
    }
    else {
        return root_2;
    }
}


double calc_volume(double h, int X, int Y) {
    return (4.0 * pow(h, 3.0)) - (2.0 * X * pow(h, 2.0)) - (2.0 * Y * pow(h, 2.0)) + (X * Y * h);
}


int main() {
    int t, X, Y;
    double h;

    cin >> t;

    cout << fixed;
    cout << setprecision(10);

    for (int i=0; i<t; i++) {
        cin >> X >> Y;

        h = calc_root_derivative(X, Y);

        cout << calc_volume(h, X, Y) << endl;
    }
}